# HW2 - Naive Bayes Classifier
##### 負責助教：陳亮宇 catchingfire80@gmail.com
 

> 這次的作業想讓大家練習一些file I/O, data preproceing, sklearn函式庫的操作

---
## 評分標準
* 上課繳交：25%
    * 在當天 16:20 前上傳至 new E3 作業，請將以下打包成 .zip 上傳：
        * 程式原始檔 (.ipynb）
        * 程式 .html 檔
* 程式分數：75%
    * 於一週內完成並上傳至 new E3

#### 注意：最終學期成績將視全班原始成績做調整

---

# 關於資料集
- 資料集: 這是一個收集了各種車子相關的數據的分類資料集，目標在藉由車子的不同特性來預測其對消費者來說的接受度。
- 屬性:
    - Class Values(target):
        - unacc, acc, good, vgood
    - Attributes(features):
        - buying: vhigh, high, med, low
        - maint: vhigh, high, med, low
        - doors: 2, 3, 4, 5more
        - persons: 2, 4, more
        - lug_boot: small, med, big
        - safety: low, med, high
- 分類任務: 分類任務旨在藉由讀入上述的features作為模型的input，期望預測出正確的分類(target)。
- 衡量模型: 為了知道自己的模型好壞，一般在拿到資料時會將資料分為training以及testing兩部分，模型訓練時僅拿training set的資料進行訓練，而模型的表現則由完全沒看過的testing set上所預測出來的準確度來衡量。

## 基本資料 - 5%

#### 姓名：Jean
#### 學號：xxxxxxxx

## Packages that Might Be Used

In [2]:
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import BernoulliNB
from sklearn.naive_bayes import MultinomialNB
#from sklearn.naive_bayes import ComplementNB
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.metrics import confusion_matrix
import numpy as np
import pandas as pd

### 相關連結

* Packages
    * [cantaloupe](https://youtu.be/zkd8bJF15pQ)
    * [numpy](http://www.numpy.org/)
    * [pandas](https://pandas.pydata.org/)
    * [re](https://docs.python.org/zh-cn/3/library/re.html)
    * [CountVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html)
    * [LabelEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html)
    * [difflib](https://docs.python.org/2/library/difflib.html)
    * [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html)
    * [Gaussian-naive-bayes](https://scikit-learn.org/stable/modules/naive_bayes.html#gaussian-naive-bayes)
    * [Multinomial-naive-bayes](https://scikit-learn.org/stable/modules/naive_bayes.html#multinomial-naive-bayes)
    * [Bernoulli-naive-bayes](https://scikit-learn.org/stable/modules/naive_bayes.html#bernoulli-naive-bayes)
    * [pyplot](https://matplotlib.org/tutorials/introductory/pyplot.html)
* Markdown
    * jupyter 可以支援 markdown 區塊，方便打上文字說明此份 notebook 在做的事情
    * [wiki](https://zh.wikipedia.org/wiki/Markdown)
    * [教學](https://markdown.tw/)

## 1. Load Data - 40%

#### 請使用 `pandas` 中的 `read_table` 讀取 car.txt - 10% - [Hint,會用到的參數有names及sep](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_table.html)

In [104]:
car = pd.read_table("car.txt", names = ['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety', 'class_value'], sep = ',')
car

buying  maint  doors persons lug_boot safety class_value
0     vhigh  vhigh      2       2    small    low       unacc
1     vhigh  vhigh      2       2    small    med       unacc
2     vhigh  vhigh      2       2    small   high       unacc
3     vhigh  vhigh      2       2      med    low       unacc
4     vhigh  vhigh      2       2      med    med       unacc
...     ...    ...    ...     ...      ...    ...         ...
1723    low    low  5more    more      med    med        good
1724    low    low  5more    more      med   high       vgood
1725    low    low  5more    more      big    low       unacc
1726    low    low  5more    more      big    med        good
1727    low    low  5more    more      big   high       vgood

[1728 rows x 7 columns]

## 分析

#### 印出 `car` 的 `前五筆資料` - 5% - [Hint](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.head.html)

In [113]:
df = pd.DataFrame(car)
print(df.head(5))

  buying  maint doors persons lug_boot safety class_value
0  vhigh  vhigh     2       2    small    low       unacc
1  vhigh  vhigh     2       2    small    med       unacc
2  vhigh  vhigh     2       2    small   high       unacc
3  vhigh  vhigh     2       2      med    low       unacc
4  vhigh  vhigh     2       2      med    med       unacc


#### 印出 `car` 的 **shape** - 5% - [hint](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shape.html)

In [106]:
print(car.shape)

(1728, 7)


> #### 可以看出 `car dataset` 具有 1728 筆資料，而每筆資料有 6 個 features 和 1 個 label(target)

#### 只印出 `car` 的 `doors` 的部分 - 5%
 * 可以google pandas, dataframe, print, column這些關鍵字~ 

In [107]:
print(car['doors'])

0           2
1           2
2           2
3           2
4           2
        ...  
1723    5more
1724    5more
1725    5more
1726    5more
1727    5more
Name: doors, Length: 1728, dtype: object


#### 印出 `car` 的 target (label) - 5%

In [108]:
print(car["class_value"])

0       unacc
1       unacc
2       unacc
3       unacc
4       unacc
        ...  
1723     good
1724    vgood
1725    unacc
1726     good
1727    vgood
Name: class_value, Length: 1728, dtype: object


#### 印出 `car` 的第 100, 150, 500, 1070 個 data instances （並包含其 target / label） - 10% - [Hint](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iloc.html)

Note: 格式可以稍微不同

In [109]:
indices = [100, 150, 500, 1070]
car.iloc[indices, :]

buying  maint  doors persons lug_boot safety class_value
100   vhigh  vhigh  5more    more    small    med       unacc
150   vhigh   high      3       4      big    low       unacc
500    high  vhigh      4       4      med   high       unacc
1070    med   high  5more       4      big   high         acc

## 2. Data Preprocessing - 30%

### Label Encoder - 15%
- 在進行以下步驟之前，助教先將dataframe轉成numpy，有利於以下的作業
    - 將 categorical(離散型資料) 進行label encoding，其功能為將離散型資料轉換為數值型態以方便後續模型處理，需要注意的是此轉換僅保留了數據的不相同部分特徵，若為處理距離的模型還需要另做轉換。
    - Encoding完之後，將 `cars` 的feature及target分開成兩個變數。

In [110]:
cars = np.array(car)
le = preprocessing.LabelEncoder()
for i in range(7):
    cars[:,i] = le.fit_transform(cars[:, i])
cars_fe = cars[:, :-1]
cars_tar = cars[:, -1]
print("feature:", cars_fe)
print("target:", cars_tar)

feature: [[3 3 0 0 2 1]
 [3 3 0 0 2 2]
 [3 3 0 0 2 0]
 ...
 [1 1 3 2 0 1]
 [1 1 3 2 0 2]
 [1 1 3 2 0 0]]
target: [2 2 2 ... 2 1 3]


### Splitting Data - 15%
 - 將資料分成training與testing set - [Hint](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html)
 - 將training data 與 testing data 以8:2的比例下去分
 - Random_state設42
 

In [111]:
X_train, X_test, y_train, y_test = train_test_split(cars_fe, cars_tar, test_size = 0.2, train_size = 0.8, random_state = 42)

## 3. Training Naive Bayes - 25% - [Hint](https://scikit-learn.org/stable/modules/naive_bayes.html)
 * 分別對 GaussianNB, BernoulliNB, MultinomialNB 用 training data 進行 training
 * 預測時使用testing data進行預測，並將預測準確率印出來(就是拿predict出來的target跟原本的target對答案)

In [112]:
gnb = GaussianNB()
clf = BernoulliNB()
mlf = MultinomialNB()
y_train = list(y_train)
y_pred_g = gnb.fit(X_train, y_train).predict(X_test)
y_pred_c = clf.fit(X_train, y_train).predict(X_test)
y_pred_m = mlf.fit(X_train, y_train).predict(X_test)
print("Gaussian:", (X_test.shape[0] - (y_test != y_pred_g).sum()) / X_test.shape[0])
print("Bernoulli:", (X_test.shape[0] - (y_test != y_pred_c).sum()) / X_test.shape[0])
print("Multinomial:", (X_test.shape[0] - (y_test != y_pred_m).sum()) / X_test.shape[0])

Number of mislabeled points out of a total 346 points : 130
Number of mislabeled points out of a total 346 points : 93
Gaussian: 0.6242774566473989
Bernoulli: 0.7312138728323699
Multinomial: 0.684971098265896


## 有沒有什麼想對助教說的話?

![](https://i.imgur.com/31NjDYX.gif?noredirect)

>實作好好玩 :)  耶呼!